### ISSUE TITLE
FPM fails to find dependency if use statement is proceeded by tab

### ISSUE THREAD

**Issue by SoupOfDepression** [fortran-lang/fpm#508]
While attempting to use FPM for the first time, I figured out that it does not recognize a use statement if you place a tab before the use statement. This can be quite confusing if you are used to using a compiler that allows you to use tabs (such as gfortran.) This is also the case for include statements.

```
program main
use helloff
end program main
```
Will compile fine (if fpm.toml is correct) while
```
program main
	use helloff
end program main
```
will give you
`Fatal Error: Cannot open module file ‘helloff.mod’ for reading at (1): No such file or directory
compilation terminated.`
no matter what you specify in fpm.toml.

This is because adjustl is used when testing for use statements, which removes spaces, but ignores tabs. (FPM also uses adjustl while testing for both use and include statements instead of right after reading the lines from the file, making the adjustl run many more times than required, so it might be a good idea to change the way it deals with indentation even if tabs don't get supported)

**Comment by milancurcic** [comment]
Thanks, I can reproduce with fpm-0.3.0. GFortran also issues a warning:

```
app/main.f90:2:1:

    2 |  use tabs, only: say_hello
      | 1
Warning: Nonconforming tab character at (1) [-Wtabs]
```

I didn't know tab was non-conforming. :)

So, I think this will be an fpm design decision whether to treat this as a bug (tabs are okay) or to improve the parsing and error message to catch this gracefully (tabs are not okay).

**Comment by awvwgk** [comment]
I'm for tabs are not okay :P

Joke aside, allowing tabs in the source code is probably easier than creating a dedicated and meaningful error for those, let's leave it to the Fortran compiler to complain about those.

Fixing the [source_parsing](https://github.com/fortran-lang/fpm/blob/master/src/fpm_source_parsing.f90] to allow tabs as whitespace could be a good first issue, if somebody wants to pick this up. From my understanding, the issue boils down to the usage of the intrinsics like `trim` and `adjustl` to strip spaces, instead we could use [`strip`](https://stdlib.fortran-lang.org/page/specs/stdlib_strings.html#strip) and [`chomp`](https://stdlib.fortran-lang.org/page/specs/stdlib_strings.html#chomp) from stdlib to strip all whitespace consistently. Also, the `delims` passed to `split_n` have to be expanded as well.

**Comment by ivan-pi** [comment]
> This can be quite confusing if you are used to using a compiler that allows you to use tabs (such as gfortran.)

Since tabs will be a source of non-portability for your fpm package I would highly recommend to set your text editor to emit spaces when pressing the tab key.

Otherwise I agree with @awvwgk, it is easier to let this fail in the compilation stage.

**Comment by urbanjost** [comment]
As tabs could cause issues in other locations in the USE,INCLUDE, MODULE lines expanding them for the Fortran input files seems appropriate, so I added the M_strings procedure notabs(3f). I did not also in the same routine do the adjustl and lower
as it would affect the hash value in a possibly significant way and prevent recompilation.  This change would also cause a source file that had tabs added/deleted to appear unchanged as well, but I cannot think of a problem with that except that some compilers might treat a tab as a space, so expanding it assuming tab stops every eight characters might produce a line that is "too long" if the compiler is called with a line length restriction (which is essentially going away as a limit in future Fortran standards, by the way), although reducing the calls to LOWER(3f) and ADJUSTL(3f) does seem like a good idea and could be done once on the entire array, with a call to READ_LINES_EXPANDED used for parsing input with everything adjusted and lowercase-ed and a call to READ_LINES done just for the hash. That would also eliminate not recompiling if the file contains new tabs or has tabs removed. But at least on my machine using the rather large general-purpose-fortran `fpm` package as a test the timing difference seemed unnoticeable, so I went for minimal changes.

**Comment by urbanjost** [comment]
PS;  I did an "fpm new A" and changed all the space character to tabs and built with ifort, gfortran, and nvfortran and ifort and nvfortran built with all the tabs and gfortran built with I changed the tab between "print" and "*"  (perhaps a bug, since it tolerates tab characters everywhere else in the simple code tested?). Since tolerating tabs seemed very common leaving it up to the compiler seemed reasonable, but the read_lines_expanded(3f) procedure could easily warn about tab characters in a compiler-independent way.  Does that seem like a good idea?

**Comment by urbanjost** [comment]
Incidental, but so no one duplicates it I reported the minor issue with a tab in a PRINT statement in gfortran 9.3.

        https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101399

**Comment by urbanjost** [comment]
For those who want to expand tab characters from the command line, note ULS have the command expand(1) as one method for removing horizontal tab characters.

**Comment by urbanjost** [comment]
At least on my machine a large build is dominated by the compiler and external system commands, but the parsing function and calls to lowercase the input files definitely does dominate the time consumed. The output of gprof for a large build follows ...

```text
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  ms/call  ms/call  name    
 61.54      0.40     0.40  3166307     0.00     0.00  __fpm_strings_MOD_lower
 13.85      0.49     0.09     1137     0.08     0.49  __fpm_source_parsing_MOD_parse_f_source
  9.23      0.55     0.06   306587     0.00     0.00  __fpm_strings_MOD_fnv_1a_char
  6.15      0.59     0.04     2263     0.02     0.02  __fpm_targets_MOD_find_module_dependency
  6.15      0.63     0.04     2259     0.02     0.02  __fpm_targets_MOD_add_target
  1.54      0.64     0.01     1225     0.01     0.01  __fpm_filesystem_MOD_read_lines
  1.54      0.65     0.01     1210     0.01     0.06  __fpm_strings_MOD_fnv_1a_string_t
  0.00      0.65     0.00    26343     0.00     0.00  __fpm_filesystem_MOD_dirname
  0.00      0.65     0.00    10167     0.00     0.00  __fpm_strings_MOD_split
  0.00      0.65     0.00     8433     0.00     0.00  __fpm_source_parsing_MOD_split_n
  0.00      0.65     0.00     7836     0.00     0.00  __fpm_filesystem_MOD_exists
  0.00      0.65     0.00     7555     0.00     0.00  __fpm_strings_MOD_string_array_contains
  0.00      0.65     0.00     6265     0.00     0.00  __fpm_strings_MOD_is_fortran_name
  0.00      0.65     0.00     5266     0.00     0.00  __fpm_strings_MOD_str_ends_with_str
  0.00      0.65     0.00     4856     0.00     0.00  __fpm_environment_MOD_get_os_type
  0.00      0.65     0.00     4649     0.00     0.00  __fpm_filesystem_MOD_canon_path
  0.00      0.65     0.00     4649     0.00     0.00  __fpm_filesystem_MOD_unix_path
  0.00      0.65     0.00     4518     0.00     0.00  __fpm_targets_MOD_add_dependency
  0.00      0.65     0.00     3805     0.00     0.00  __fpm_strings_MOD_str_ends_with_any
```

**Comment by LKedward** [comment]
I assume you're on a non-windows machine then @urbanjost - yes outside of Windows, the source parsing is very much the dominating process. I'm working on a refactor to improve this slightly.

**Comment by urbanjost** [comment]
OK. I was going to refactor it a bit to reduce the calls to lower and a few other things but if you are already looking at that I will not. I have a MSWindows machine now, but generally only use it to check the portability of software developed in ULS platforms.

**Comment by urbanjost** [comment]
PS:  A simplistic approach to reducing the number of calls to lower I had already tried where I just made a duplicate of file_lines as soon as file_lines was created because  I needed INCLUDE #include lines to retain their case because of file names did show a drop from 0.69 to 0.27 seconds on an "fpm build" when everything was already up to date, which was significant on a large build like general-purpose fortran. The same run as above with just that change produced
```text
 Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  ms/call  ms/call  name    
 22.22      0.06     0.06     1137     0.05     0.14  __fpm_source_parsing_MOD_parse_f_source
 20.37      0.12     0.06   306587     0.00     0.00  __fpm_strings_MOD_fnv_1a_char
 18.52      0.17     0.05     2263     0.02     0.02  __fpm_targets_MOD_find_module_dependency
 14.81      0.21     0.04   333783     0.00     0.00  __fpm_strings_MOD_lower
 14.81      0.25     0.04     2259     0.02     0.02  __fpm_targets_MOD_add_target
  3.70      0.26     0.01    10167     0.00     0.00  __fpm_strings_MOD_split
  3.70      0.27     0.01     7555     0.00     0.00  __fpm_strings_MOD_string_array_contains
  1.85      0.27     0.01     1210     0.00     0.05  __fpm_strings_MOD_fnv_1a_string_t
  0.00      0.27     0.00    23443     0.00     0.00  __fpm_filesystem_MOD_dirname
  0.00      0.27     0.00     8433     0.00     0.00  __fpm_source_parsing_MOD_split_n
  0.00      0.27     0.00     6265     0.00     0.00  __fpm_strings_MOD_is_fortran_name
  0.00      0.27     0.00     5577     0.00     0.00  __fpm_filesystem_MOD_exists
```
that is a pretty big build so probably not typical, but it was a noticeable change.

**Comment by LKedward** [comment]
> I was going to refactor it a bit to reduce the calls to lower and a few other things but if you are already looking at that I will not.

Please do go ahead and submit your patch to improve this @urbanjost! (My refactoring is at a higher level to restructure the routines and allow other issues to be addressed, so I don't think there'll be any conflict.)

**Comment by urbanjost** [comment]
A brute force approach did bring the number of calls to lower(3f)  down by a factor of ten from 3M to 300 000, and results in a little over a 8.5/100 reduction in time in the fpm routines, so thanks to @SoupOfDepression for mentioning that!  If the PR is accepted that actually produces a message and enforces the name only allowed to be an alphanumeric starting with a letter except for allowing hypens and dashes I'll update the documentation too;  although hopefully getting the message when trying to use a non-conforming name would have prevented confusion.  I had a similiar issue in the past when I could not tell if names for executables could be pathnames that overrode being placed in the build directory quite a while ago, especially because it "worked". Now I know how that would not be a good "feature" if you want to keep different profile builds separate (and would have security issues if it allowed full pathnames, but the names were treated as relative to the package top).

**Comment by awvwgk** [comment]
Fixed by #521
